### PR TITLE
Added reshape, reshape2, squeeze and squeeze2 BF16/FP32 FWD/BWD kernels

### PR DESCRIPTION
**Pull request by jakpiase**
### PR types
New features

### PR changes
OPs

### Describe
Added reshape, reshape 2, squeeze and squeeze2 BF16/FP32 FWD/BWD kernels for enabling word2vec bf16 training.

### CODE REVIEW

File: paddle/fluid/operators/mkldnn/squeeze_mkldnn_op.cc [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+limitations under the License. */
+
+#include "paddle/fluid/operators/squeeze_op.h"
+#include "paddle/fluid/platform/mkldnn_reuse.h"
+
+namespace paddle {
+namespace operators {
+
+framework::DDim GetOutputShape(const std::vector<int> squeeze_dims,
+                               const framework::DDim& in_dims, bool is_runtime);
```
Comment (arogowie-intel) [inline paddle/fluid/operators/mkldnn/squeeze_mkldnn_op.cc]: Is this forward declaration needed? This function signature is already available in `squeeze_op.h` header.

File: paddle/fluid/operators/mkldnn/squeeze_mkldnn_op.cc [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+    auto* out = ctx.Output<LoDTensor>("Out");
+
+    auto& axes = ctx.Attr<std::vector<int>>("axes");
+    auto x_dims = x->dims();
+    auto x_vec_dims = framework::vectorize(x_dims);
+    auto out_dims = GetOutputShape(axes, x_dims, true);
+
+    mkldnn::memory::data_type x_type = framework::ToMKLDNNDataType(x->type());
+    std::string key = platform::CreateKey(dev_ctx, x_vec_dims, x->format(),
+                                          x->format(), x_type);
```
Comment (arogowie-intel) [inline paddle/fluid/operators/mkldnn/squeeze_mkldnn_op.cc]: Is it correct that you create the key with doubled `x->format()` ?

File: paddle/fluid/operators/mkldnn/squeeze_mkldnn_op.cc [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+                   ops::SqueezeMKLDNNKernel<float>,
+                   ops::SqueezeMKLDNNKernel<paddle::platform::bfloat16>);
+
+REGISTER_OP_KERNEL(squeeze_grad, MKLDNN, paddle::platform::CPUPlace,
+                   ops::SqueezeGradMKLDNNKernel<float>,
+                   ops::SqueezeGradMKLDNNKernel<paddle::platform::bfloat16>);
+
+REGISTER_OP_KERNEL(squeeze2, MKLDNN, paddle::platform::CPUPlace,
+                   ops::SqueezeMKLDNNKernel<float>,
+                   ops::SqueezeMKLDNNKernel<paddle::platform::bfloat16>);
```
Comment (arogowie-intel) [inline paddle/fluid/operators/mkldnn/squeeze_mkldnn_op.cc]: What is the difference between `squeeze` and `squeeze2` operators?

File: python/paddle/fluid/tests/unittests/mkldnn/test_squeeze2_mkldnn_op.py [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+    def setUp(self):
+        self.set_op_type()
+        self.init_test_case()
+        self.x = np.random.random(self.ori_shape).astype("float32")
+        self.set_inputs()
+        self.init_attrs()
+        self.set_outputs()
+
+    def test_check_output(self):
+        self.check_output_with_place(core.CPUPlace(), no_check_set=['XShape'])
```
Comment (arogowie-intel) [inline python/paddle/fluid/tests/unittests/mkldnn/test_squeeze2_mkldnn_op.py:56]: Dose setting `no_check_set=['XShape']` will not prevent checking for this output existence?

File: paddle/fluid/operators/mkldnn/squeeze_mkldnn_op.cc (hunk 1 above)
Comment (jakpiase) [reply paddle/fluid/operators/mkldnn/squeeze_mkldnn_op.cc]: You're right, I'll delete that

File: paddle/fluid/operators/mkldnn/squeeze_mkldnn_op.cc (hunk 2 above)
Comment (jakpiase) [reply paddle/fluid/operators/mkldnn/squeeze_mkldnn_op.cc]: Yes, this reorder should be done without changing memory format

File: paddle/fluid/operators/mkldnn/squeeze_mkldnn_op.cc (hunk 3 above)
Comment (jakpiase) [reply paddle/fluid/operators/mkldnn/squeeze_mkldnn_op.cc]: squeeze has "X" input available in backward propagation and I can get input dims from here while squeeze2 has different tensor made just to transfer input dims to backward propagation(XShape tensor, its only purpose is to transfer dims between forward and backward pass)

File: python/paddle/fluid/tests/unittests/mkldnn/test_squeeze2_mkldnn_op.py (hunk 4 above)
Comment (jakpiase) [reply python/paddle/fluid/tests/unittests/mkldnn/test_squeeze2_mkldnn_op.py:56]: XShape output is irrelevant in this op. Its only purpose is to transfer X tensor's dims between forward and backward pass

**Review by jczaja — DISMISSED** [review]

**Review by jczaja — DISMISSED** [review]

**Review by jczaja — APPROVED** [review]

**Review by lidanqing-vv — APPROVED** [review]
LGTM.

**Pull request merged**